### PR TITLE
VITIS-10045 - Merge 2 patchers

### PR DESCRIPTION
**Pull request by dezhiAmd**
#### Problem solved by the commit
[PR7790](https://github.com/Xilinx/XRT/pull/7790) added a patcher to patch control-packet for IPU despite there is already a patcher for TurinAI

#### How problem was solved, alternative solutions (if any) and why they were rejected
Merge the patcher for IPU with the patcher for TurinAI

#### What has been tested and how, request additional testing if necessary
Test on IPU is done using the test application xrt_flow_test. Test on TurinAI is done using shimTest 26 on simnow

### CODE REVIEW

**Review by stsoe — APPROVED** [review]
Nice. LGTM.

**Review by stsoe — CHANGES_REQUESTED** [review]

File: src/runtime_src/core/common/api/xrt_module.cpp [hunk 1]
```diff
@@ -111,15 +110,13 @@ struct patchers
...
 
   // Offsets from base address of control code buffer object
   // The base address is passed in as a parameter to patch()
   std::vector<uint64_t> m_ctrlcode_offset;
 
-  patchers(symbol_type type, std::string name, std::vector<uint64_t> ctrlcode_offset)
+  patcher(symbol_type type, const std::vector<uint64_t>& ctrlcode_offset)
     : m_symbol_type(type)
-    , m_buf_name(name)
     , m_ctrlcode_offset(ctrlcode_offset)
```
[the remark this answers is not in the document]
Comment (dezhiAmd) [reply src/runtime_src/core/common/api/xrt_module.cpp:120]: You are correct. But I realize mbufname is not being used and it is removed in this commit.

File: src/runtime_src/core/common/api/xrt_module.cpp [hunk 2]
```diff
@@ -716,12 +602,19 @@ class module_elf : public module_impl
...
     , m_os_abi{ xrt_core::elf_int::get_elfio(m_elf).get_os_abi() }
-    , m_ctrlcodes{ initialize_column_ctrlcode(xrt_core::elf_int::get_elfio(m_elf)) }
-    , m_arg2patcher{ initialize_arg_patchers(xrt_core::elf_int::get_elfio(m_elf), m_ctrlcodes) }
-    , m_instr_buf{ initialize_instr_buf(xrt_core::elf_int::get_elfio(m_elf)) }
-    , m_ctrl_packet{ initialize_ctrl_packet(xrt_core::elf_int::get_elfio(m_elf)) }
-    , m_arg2patchers{ initialize_arg_patchers(xrt_core::elf_int::get_elfio(m_elf), m_instr_buf, m_ctrl_packet) }
-  {}
+  {
+    if ( m_os_abi == Elf_Amd_Aie2ps)
+    {
```
Comment (larry9523) [inline src/runtime_src/core/common/api/xrt_module.cpp:607]: if (m_os_abi == Elf_Amd_Aie2ps) {

File: src/runtime_src/core/common/api/xrt_module.cpp [hunk 3]
```diff
@@ -716,12 +602,19 @@ class module_elf : public module_impl
...
+      m_ctrlcodes = initialize_column_ctrlcode(xrt_core::elf_int::get_elfio(m_elf));
+      m_arg2patcher = initialize_arg_patchers(xrt_core::elf_int::get_elfio(m_elf), m_ctrlcodes);
+    }
+    else if ( m_os_abi == Elf_Amd_Aie2p)
+    {
```
Comment (larry9523) [inline src/runtime_src/core/common/api/xrt_module.cpp:612]: } else if (m_os_abi == Elf_Amd_Aie2p) {

File: src/runtime_src/core/common/api/xrt_module.cpp [hunk 4]
```diff
@@ -982,15 +875,15 @@ class module_sram : public module_impl
...
+    if (os_abi == Elf_Amd_Aie2ps) {
       if (m_patched_args.size() != m_parent->number_of_arg_patchers()) {
         auto fmt = boost::format("ctrlcode requires %d patched arguments, but only %d are patched")
             % m_parent->number_of_arg_patchers() % m_patched_args.size();
         throw std::runtime_error{ fmt.str() };
       }
       m_buffer.sync(XCL_BO_SYNC_BO_TO_DEVICE);
     }
-    else {
+    else if (os_abi == Elf_Amd_Aie2p) {
```
Comment (larry9523) [inline src/runtime_src/core/common/api/xrt_module.cpp:886]: } else if (os_abi == Elf_Amd_Aie2p) {

File: src/runtime_src/core/common/api/xrt_module.cpp [hunk 5]
```diff
@@ -1009,13 +902,12 @@ class module_sram : public module_impl
...
       create_instr_buf(m_parent.get());
       create_ctrlpkt_buf(m_parent.get());
-
       fill_bo_addresses();
-      return;
     }
-
-    create_instruction_buffer(m_parent.get());
-    fill_column_bo_address(m_parent->get_data());
+    else if (os_abi == Elf_Amd_Aie2ps) {
```
Comment (larry9523) [inline src/runtime_src/core/common/api/xrt_module.cpp:907]: } else if (os_abi == Elf_Amd_Aie2ps) {

**Review by larry9523 — APPROVED** [review]
Overall looks good to me. Just a few coding style nits.

File: src/runtime_src/core/common/api/xrt_module.cpp (hunk 3 above)
Comment (stsoe) [reply src/runtime_src/core/common/api/xrt_module.cpp:612]: It's important  to me (actually not really) that `else` lines up with `if`.

```
if (cond1) {
  stmt1;
  stmt2;
}
else if (cond 2) {
  stmt3;
  stmt4
}  
```

File: src/runtime_src/core/common/api/xrt_module.cpp (hunk 4 above)
Comment (stsoe) [reply src/runtime_src/core/common/api/xrt_module.cpp:886]: Ditto :-)

File: src/runtime_src/core/common/api/xrt_module.cpp (hunk 5 above)
Comment (stsoe) [reply src/runtime_src/core/common/api/xrt_module.cpp:907]: Ditto :-)

File: src/runtime_src/core/common/api/xrt_module.cpp (hunk 3 above)
Comment (larry9523) [reply src/runtime_src/core/common/api/xrt_module.cpp:612]: So user space code we follow this rule but in kernel space (Linuxe), we put "}" and "else/ elseif" in same line?

Comment (stsoe) [reply src/runtime_src/core/common/api/xrt_module.cpp:612]: I am sure there are many conventions that are different between userspace and kernel space :-) IMO is it much easier to scan code when things are lined up to the left, so this is the convention I follow in userspace.

Comment (larry9523) [reply src/runtime_src/core/common/api/xrt_module.cpp:612]: No problem. I just did not know this before. I will follow this rule when writing user space code.  Thanks.

**Review by larry9523 — APPROVED** [review]

File: src/runtime_src/core/common/api/xrt_module.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (stsoe) [reply src/runtime_src/core/common/api/xrt_module.cpp:120]: You missed the vector

[the remark this answers is not in the document]
Comment (dezhiAmd) [reply src/runtime_src/core/common/api/xrt_module.cpp]: Fixed

**Review by stsoe — CHANGES_REQUESTED** [review]
I have one request related to initialize_ctrl_packet(...).
In that function a variable ctrldata initialized to nullptr and a loop searches for .ctrldata section which is then assigned to outer variable. 
What if no .ctlrdata section was found, then apppend_section_data will crash. Isn't it better to call append within the loop when the .ctrldata section is actually found?
Either that or an exception.

File: src/runtime_src/core/common/api/xrt_module.cpp [hunk 6]
```diff
@@ -428,25 +348,19 @@ class module_elf : public module_impl
   instr_buf
   initialize_instr_buf(const ELFIO::elfio& elf)
   {
-    if (m_os_abi != Elf_Amd_Aie2p) {
-      XRT_PRINTF("module_elf::initialize_instr_buf(), not AIE2/AIE2P, skip...\n");
-      return {};
-    }
-
+    instr_buf instrbuf;
     ELFIO::section* ctrltext = nullptr;
```
Comment (stsoe) [inline src/runtime_src/core/common/api/xrt_module.cpp:352]:
```suggestion
```

File: src/runtime_src/core/common/api/xrt_module.cpp [hunk 7]
```diff
@@ -428,25 +348,19 @@ class module_elf : public module_impl
...
 
     for (const auto& sec : elf.sections) {
       auto name = sec->get_name();
       // Instruction buffer is in .ctrltext section.
       if (name.find(".ctrltext") != std::string::npos) {
         ctrltext = sec.get();
+        instrbuf.append_section_data(ctrltext);
```
Comment (stsoe) [inline src/runtime_src/core/common/api/xrt_module.cpp:359]:
```suggestion
        instrbuf.append_section_data(sec.get());
```

File: src/runtime_src/core/common/api/xrt_module.cpp [hunk 8]
```diff
@@ -455,25 +369,19 @@ class module_elf : public module_impl
   control_packet
   initialize_ctrl_packet(const ELFIO::elfio& elf)
   {
-    if (m_os_abi != Elf_Amd_Aie2p) {
-      XRT_PRINTF("module_elf::initialize_ctrl_packet(), not AIE2/AIE2P, skip...\n");
-      return {};
-    }
-
+    control_packet ctrlpacket;
     ELFIO::section* ctrldata = nullptr;
```
Comment (stsoe) [inline src/runtime_src/core/common/api/xrt_module.cpp:373]:
```suggestion
```

File: src/runtime_src/core/common/api/xrt_module.cpp [hunk 9]
```diff
@@ -455,25 +369,19 @@ class module_elf : public module_impl
...
 
     for (const auto& sec : elf.sections) {
       auto name = sec->get_name();
-      // Control Packet is in .ctrldata section.
+      // Control Packet is in .ctrldata section
       if (name.find(".ctrldata") != std::string::npos) {
         ctrldata = sec.get();
+        ctrlpacket.append_section_data(ctrldata);
```
Comment (stsoe) [inline src/runtime_src/core/common/api/xrt_module.cpp:380]:
```suggestion
        ctrlpacket.append_section_data(sec.get());
```

**Review by stsoe — CHANGES_REQUESTED** [review]

**Review by stsoe — APPROVED** [review]
Looks good.  Thank you!

**Pull request merged**